torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -1056,6 +1008,74 @@ struct GraphFuser {
     aliasDb_ = AliasAnalysis(graph_);
   }
 
+  bool canFuseWithConcat(Value* producer, Node* before_check) {
+    if (!isFusable(producer->node())) {
+      return false;
+    }
+    // NB: it is important that this check happens after isFusable, which checks
+    // that the blocks match, and it's not a special node like prim::Param
+    if (!producer->node()->couldMoveBeforeTopologically(
+            before_check, aliasDb_.value())) {
+      return false;
+    }
+    // Fusion groups can be merged with concat's group if and only if
+    // the value they produce isn't already coming from a concat.
+    if (producer->node()->kind() == prim::FusionGroup) {
+      auto subgraph = producer->node()->g(attr::Subgraph);
+      auto* node = subgraph->outputs().at(producer->offset())->node();
+      return node->kind() != prim::FusedConcat;
+    }
+    return true;
+  }
+
+  Node* createFusedConcat(Node* node) {
+    JIT_ASSERT(node->kind() == aten::cat);
+
+    Graph* graph = node->owningGraph();
+    Node* list_construct = node->namedInput(attr::tensors)->node();
+    int64_t dim = node->get<int64_t>(attr::dim).value();
+
+    Node* fused_cat = graph->create(prim::FusedConcat, list_construct->inputs())
+                          ->i_(attr::dim, dim);
+    fused_cat->insertBefore(list_construct);
+    fused_cat->output()->copyMetadata(node->output());
+    node->output()->replaceAllUsesWith(fused_cat->output());
+
+    // NB: this deletes the fused_cat node from the original graph
+    return createSingletonFusionGroup(fused_cat);
+  }
+
+  void fuseConcats() {
+    for (auto it = block_->nodes().rbegin(); it != block_->nodes().rend();
+         ++it) {
+      Node* cat = *it;
+      if (!isFusableCatNode(cat)) {
+        continue;
+      }
+      Node* list_construct = cat->namedInput(attr::tensors)->node();
+      Node* fused_cat = createFusedConcat(cat);
```

Comment:
Does this mean we can hypothetically have singleton FusionGroups that only contain a `FusedConcat` node?

Reply:
Yes. I was actually going to profile this, but I would really expect that a specialized kernel for concat like this one (with fixed number of arguments) will beat the stock one.

Reply:
Ok seems like the default one is _slightly_ faster. I will inline fused concats back.

Reply:
It's an occupancy problem, generally - stock concat has a thread per output element, give or take, generated concat kernel has a thread per single input tensor element, so every thread iterates over inputs and outputs. Stock concat can saturate device faster. Also, generated kernel has an insane number of arguments which probably adds to launch overhead, stock concat has h2d copy to pass all the TensorInfo's, which is also an overhead, but may be a smaller one?